### ISSUE TITLE
Set BPM value

### ISSUE THREAD

**Issue by thiago-s-medeiros** [fmalcher/soundcraft-ui#148]
Hi there!

I've downloaded the repo like 10min ago but I'm missing the BPM set value... would that be possible?

when I use thru the UI it sends always this 3 values at the same time... I think this should be the same here...

![Captura de Tela 2024-10-17 às 15 57 00](https://github.com/user-attachments/assets/7fb65ded-25f9-4cc9-9f87-896ceb46c0fc)

**Comment by fmalcher** [comment]
Sure, I think this is possible. However, we need to clear up a few things before:
- What hardware model do you have? We need to find out whether there are differences between the three models.
- Is there any other place where the BPM info is necessary except for the Delay FX 3? The TAP button over there only sets BPM for `f.2`, not globally.
- I'm not sure where this should be placed in the library API. There is no section for FX settings yet (cc #140) and I'm reluctant to create a whole new controller class just for the BPM setting.


If anybody has some input here, and we can clear up all these things, I can easily implement this.

**Comment by thiago-s-medeiros** [comment]
So...
1 is the reverb/delay (second effect), 2 is the delay/Chorus (third effect) and 3 is the room/delay (fourth effect)

I've created a method and a variable in the master-bus just to test it:

```
bpm$ = this.store.state$.pipe(select(selectBPM(1))); //I've selected the 1 but could be 1, 2 or 3... I haven't found the global bpm

setBPM(value: string) {
    this.conn.sendMessage(`SETD^f.1.bpm^${value}`);
    this.conn.sendMessage(`SETD^f.2.bpm^${value}`);
    this.conn.sendMessage(`SETD^f.3.bpm^${value}`);
  }
  ```
  
a selector in `state-selectors` like this:
  ```
  export const selectBPM: Selector<number> = (channel: 1 | 2 | 3) => state =>
  getValueFromObject<number>(state, `f.${channel}.bpm`);
  ```
  
 
  
  and in the HTML is like this:
  ```
  <div>
  <label>BPM</label>
  <input
    type="number"
    (input)="master.setBPM(BpmInput.value)"
    [value]="master.bpm$ | async"
    #BpmInput
  />
</div>
```

I'm sure that could be better but just to test if works and yes, it is possible, if anyone need this, there it is!

**Comment by thiago-s-medeiros** [comment]
Oh, hi @fmalcher , I didn't think you would anwser so fast, haha

- I'm using a ui24r.
- From my tests, the last bpm that I change in the fx settings, is the bpm that shows in the global tap in the main screen (in the UI24r ui), I will test if this becomes the global bpm. But when the global bpm is true on 1, 2 and 3... the global tap bpm changes all of them
- My idea is just to change the bpm when something else change this... if I play 3 songs, I could change the bpm (for the fx) to be the same as the music.

**Comment by thiago-s-medeiros** [comment]
Since the `f.0.bpm` isn't useful for the effect itself (the reverb), I think we could get the `f.0.bpm` as the `global bpm`, because it is always with `global bpm` true... 🤔

**Comment by thiago-s-medeiros** [comment]
oh... now I did understand the 
> is local to the browser, so we cannot work with this setting.

I can't get which effect has global tap on... but with that said, If I create a `global tap` toggle in the UI of my app, I could do the same as the ui24r do... change the BPM on the effects that have the `global tap` on

**Comment by fmalcher** [comment]
From what I can observe, `f.0.bpm` never changes its value and is always set to `120`. The global "TAP TEMPO" button sets `f.1.bpm`, `f.2.bpm` and `f.3.bpm` – even though FX `f.1` and `f.3` do not have any BPM setting. I think we can ignore `f.0`.

My biggest concern remains: We have to keep later changes in the API surface as small as possible, so this needs a proper API design. Where is a good place in the library API to control the BPM setting? At first sight, this is somewhere around the FX settings. however, we don't have any API for FX settings yet.

**Comment by thiago-s-medeiros** [comment]
The `f.1` and `f.3` could have bpm setting if you change the `f.1` from `Reverb` to `Delay` and `f.3` from `room` to `delay`.

**Comment by fmalcher** [comment]
> The f.1 and f.3 could have bpm setting if you change the f.1 from Reverb to Delay and f.3 from room to delay.

How can you do this?

**Comment by thiago-s-medeiros** [comment]
the idea of using `f.0` as global is because if you `SET` the value of bpm in it, it will change, and we could use this value as the "global" BPM... when I toggle some effect to `global tap` (inside my app), it could get this value and set in the others effects selected as `global tap = true`.

but this could be just a bad idea 😰

**Comment by thiago-s-medeiros** [comment]
> How can you do this?

Just long press at the name of the effect and it will open a box to select the effect

https://github.com/user-attachments/assets/cbc89be1-8bba-447d-918c-bed34b8b09e7

**Comment by fmalcher** [comment]
> Just long press at the name of the effect and it will open a box to select the effect

Oh wow, I did not know this, haha 😄 thanks!

**Comment by fmalcher** [comment]
OK, so my first idea would be to add this to the `FxBus` class (`conn.fx(1)`). Since all FX can have a BPM setting (regardless of whether it will be used), I think it is okay to add this globally.

**Comment by fmalcher** [comment]
Done and released in [2.6.0](https://github.com/fmalcher/soundcraft-ui/releases/tag/v2.6.0)! 🥳 
If you want to support my work, please consider using GitHub Sponsors to buy me a coffee: https://github.com/sponsors/fmalcher ☕
